src/groups/mqb/mqbnet/mqbnet_channel.cpp

```diff
@@ -842,45 +844,67 @@ void Channel::threadFn()
                 if (d_buffer.popFront(&item) == 0) {
                     BSLS_ASSERT_SAFE(item);
                     mode = e_PRIMARY;
+                    BALL_LOG_TRACE << "'" << d_description
+                                   << "': going to PRIMARY mode";
                 }
             } break;
             case e_PRIMARY: {
                 if (d_buffer.tryPopFront(&item)) {
-                    // Primary is empty.  Flush builders and drain secondary.
-
-                    mode = e_IDLE;
+                    // Primary is empty.
+                    mode = e_IDLE_PRIMARY;
+                    BALL_LOG_TRACE << "'" << d_description
+                                   << "': going to IDLE_PRIMARY mode";
                 }
                 else if (item->isSecondary()) {
                     // Instead of writing immediately, push it to the secondary
                     // buffer which will get flushed later.
-                    // This is done to reduce the rate of 'compactable' items
-                    // such as Replication Receipts which can accumulate
-                    // multiple receipts in one item.
+                    // This is done for ACK messages to add back pressure on
+                    // producers.
                     d_secondaryBuffer.pushBack(
                         bslmf::MovableRefUtil::move(item));
                     item.reset();
                 }
             } break;
-            case e_IDLE: {
-                // Idle.  First, flush all builders.
+            case e_IDLE_PRIMARY: {
+                // Primary buffer is exhausted, flush high priority items that
+                // are written to message builders.
                 if (flushAll(channel) == bmqio::StatusCategory::e_SUCCESS) {
-                    // Then, drain the secondary buffer
+                    // Then, drain the secondary buffer.
                     mode = e_SECONDARY;
+                    BALL_LOG_TRACE << "'" << d_description
+                                   << "': going to SECONDARY mode";
                 }
             } break;
             case e_SECONDARY: {
                 if (d_secondaryBuffer.tryPopFront(&item)) {
```

Comment:
Also, we only use `d_secondaryBuffer` from one thread. It is a thread safe `MonitoredQueue`, and we can use a standard non thread safe queue or vector here to make it faster.